base/compiler/abstractinterpretation.jl

```diff
@@ -788,20 +790,24 @@ end
 function abstract_call_builtin(interp::AbstractInterpreter, f::Builtin, fargs::Union{Nothing,Vector{Any}},
         argtypes::Vector{Any}, sv::InferenceState, max_methods::Int)
     la = length(argtypes)
-    if f === ifelse && fargs isa Vector{Any} && la == 4 && argtypes[2] isa Conditional
-        # try to simulate this as a real conditional (`cnd ? x : y`), so that the penalty for using `ifelse` instead isn't too high
-        cnd = argtypes[2]::Conditional
-        tx = argtypes[3]
-        ty = argtypes[4]
-        a = ssa_def_slot(fargs[3], sv)
-        b = ssa_def_slot(fargs[4], sv)
-        if isa(a, Slot) && slot_id(cnd.var) == slot_id(a)
-            tx = typeintersect(tx, cnd.vtype)
-        end
-        if isa(b, Slot) && slot_id(cnd.var) == slot_id(b)
-            ty = typeintersect(ty, cnd.elsetype)
-        end
-        return tmerge(tx, ty)
+    if f === ifelse && fargs isa Vector{Any} && la == 4 && (cnd = argtypes[2]; isa(cnd, Conditional))
+        if (newcnd = widenconditional(cnd); isa(newcnd, Const))
```

Comment:
style nit: we try to avoid nesting statements, particularly assignments
```suggestion
        newcnd = widenconditional(cnd)
        if isa(newcnd, Const)
```